server/src/internalClusterTest/java/org/opensearch/clustermanager/ClusterManagerTaskThrottlingIT.java

```diff
@@ -0,0 +1,214 @@
+                    if (action.contains("mapping")) {
+                        throttledRequest.incrementAndGet();
+                        assertEquals(ClusterManagerThrottlingException.class, error.getClass());
+                    }
+                }
+            });
+
+            ActionListener listener = new ActionListener() {
+                @Override
+                public void onResponse(Object o) {
+                    latch.countDown();
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    latch.countDown();
+                    throw new AssertionError(e);
+                }
+            };
+
+            executePutMappingRequests(totalRequest, dataNode, listener);
+            latch.await();
+
+            assertEquals(totalRequest + throttledRequest.get(), requestCountOnMaster.get());
+            assertBusy(
+                () -> { assertEquals(clusterService().getMasterService().numberOfThrottledPendingTasks(), throttledRequest.get()); }
+            );
+        } finally {
+            clusterSettingCleanUp();
+        }
+    }
+
+    /*
+     * This will test the throttling feature for single node.
+     *
+     * Here we will assert the client behaviour that client's request is not
+     * failed, i.e. Throttling exception is not passed to the client.
+     * Data node will internally do the retry and request should pass.
+     *
+     */
+    public void testThrottlingForSingleNode() throws Exception {
+        try {
+            internalCluster().beforeTest(random());
+            String node = internalCluster().startNode();
+            int throttlingLimit = randomIntBetween(1, 5);
+            createIndex("test");
+            setPutMappingThrottlingLimit(throttlingLimit);
+
+            AtomicInteger successfulRequest = new AtomicInteger();
+            int totalRequest = randomIntBetween(throttlingLimit, 3 * throttlingLimit);
+            CountDownLatch latch = new CountDownLatch(totalRequest);
+
+            ActionListener listener = new ActionListener() {
+                @Override
+                public void onResponse(Object o) {
+                    latch.countDown();
+                    successfulRequest.incrementAndGet();
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    latch.countDown();
+                    throw new AssertionError(e);
+                }
+            };
+            executePutMappingRequests(totalRequest, node, listener);
+
+            latch.await();
+            assertEquals(totalRequest, successfulRequest.get());
+        } finally {
+            clusterSettingCleanUp();
+        }
+    }
+
+    /*
+     * This will test the timeout of tasks during throttling.
+     *
+     * Here we will assert the client behaviour that client's request is not
+     * failed with throttling exception but timeout exception.
```

Comment:
can you pls help me understand why this is the right behavior ?

Reply:
This throttling is introduced for internal task submission, cx don't have to be aware of it. 

**Without throttling**, such tasks will wait in master's pending task queue. If tasks gets timeout in queue itself, then master return `ClusterProcessEventTimeoutException`. 

**With throttling** we are making those tasks wait on data node itself instead of master node. So let's say tasks was throttled throughout the time and gets timedout, it is the same as task is waiting in master queue without throttling. 

From cx perspective, for both case master node is busy and not able to process the event in given timeout. For that we should throw timeout exception only.

Reply:
I would prefer `Throttling` exception as the task didn't wait in master's pending task queue , but was outright rejected . The customers which are submitting the task should be aware of the difference between two . This would help us in the future if we want to take different actions on each of those.  What do you think ?

Reply:
As of now we can keep it as internal, we will go incremental and expose this to cx later. As we will need to expose various retries configuration as well for tuning. Till then cx can treat it as timeout exception only. 

I will create followup sim for this.

Reply:
+1 lets create a different SIM to expose ThrottlingException to user. Also need to cross check if this would require changes in the clients

Reply:
Created followup SIM for it: https://github.com/opensearch-project/OpenSearch/issues/4724